internal/aws/xray/telemetry/recorder.go

```diff
@@ -0,0 +1,375 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/xray/telemetry"
+
+import (
+	"errors"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/xray"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/awsutil"
+	awsxray "github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/xray"
+)
+
+const (
+	envAWSHostname   = "AWS_HOSTNAME"
+	envAWSInstanceID = "AWS_INSTANCE_ID"
+
+	queueSize       = 30
+	batchSize       = 10
+	defaultInterval = time.Minute
+)
+
+type Recorder interface {
+	// Start send loop.
+	Start()
+	// Stop send loop.
+	Stop()
+	// RecordSegmentsReceived adds the count to the record for the current period.
+	RecordSegmentsReceived(count int)
+	// RecordSegmentsSent adds the count to the record for the current period.
+	RecordSegmentsSent(count int)
+	// RecordSegmentsSpillover adds the count to the record for the current period.
+	RecordSegmentsSpillover(count int)
+	// RecordSegmentsRejected adds the count to the record for the current period.
+	RecordSegmentsRejected(count int)
+	// RecordConnectionError categorizes the error and increments the count by one for the current period.
+	RecordConnectionError(err error)
+}
+
+type RecorderOption interface {
+	apply(r *telemetryRecorder)
+}
+
+type recorderOptionFunc func(r *telemetryRecorder)
+
+func (ro recorderOptionFunc) apply(r *telemetryRecorder) {
+	ro(r)
+}
+
+func WithResourceARN(resourceARN string) RecorderOption {
+	return recorderOptionFunc(func(r *telemetryRecorder) {
+		r.resourceARN = resourceARN
+	})
+}
+
+func WithInstanceID(instanceID string) RecorderOption {
+	return recorderOptionFunc(func(r *telemetryRecorder) {
+		r.instanceID = instanceID
+	})
+}
+
+func WithHostname(hostname string) RecorderOption {
+	return recorderOptionFunc(func(r *telemetryRecorder) {
+		r.hostname = hostname
+	})
+}
+
+func WithLogger(logger *zap.Logger) RecorderOption {
+	return recorderOptionFunc(func(r *telemetryRecorder) {
+		r.logger = logger
+	})
+}
+
+// ToRecorderOptions gets the metadata recorder options if enabled by the config.
+func ToRecorderOptions(cfg Config, sess *session.Session, settings *awsutil.AWSSessionSettings) []RecorderOption {
+	if !cfg.IncludeMetadata {
+		return nil
+	}
+	hostname := os.Getenv(envAWSHostname)
+	instanceID := os.Getenv(envAWSInstanceID)
```

Comment:
Is it possible to have these loaded via the passed config?

Reply:
Added optional fields in the telemetry config.